### PR TITLE
Issue/4401 direct reply notif

### LINKED ISSUE #4401: Direct reply to comments on notifications (Android N)
Implement inline reply for comments on Android N

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #4401 

I'm chaining this PR to #4667 as we need the service to implement this one.

notif_direct_reply

To test:
1. Log in to the Android WordPress app on an Android 7.x device or emulator
2. Have someone like/comment on a blog post of yours
3. See the notification arrive at your device
4. Click on the "REPLY" quick action
5. the edittext appears where you can type the reply
6. tap the SEND button
7. watch the comment being sent and check the comment arrives at the other end (i.e. is received by the first commenter)

cc @kwonye

### CODE REVIEW

**Comment by mzorz** [comment]
I just added support for the following situations on this commit  https://github.com/wordpress-mobile/WordPress-Android/pull/4673/commits/a86dd30dd1921a0681e0bf3477c9426c2780efc5

On Android N:
-> voice Reply (wearable) works directly by invoking the service
-> reply has the direct-reply feature
-> likes and approvals invoke the service and performs the operation in the background, informing the user

On Android M and below, as we need to invoke an activity (otherwise no UI is rendered):
-> voice Reply (wearable) is handled by the activity which “traps” the reply converted text and invokes the service to complete the action of replying seamlessly
-> tapping reply invokes the activity (shows the comment reply activity)
-> likes and approvals invoke the service directly as on Android N.

If we didn’t make this differentiation, when the screen is locked there would be no hints that the “reply” button actually did anything on Android 6.0 and below (as a service was invoked and then an activity was started from there, which has no consequence in asking the user for a screen unlock)

**Comment by drw158** [comment]
I agree with with @daniloercoli that the UX should be optimistic with the Like action. I am imagining something like:

- tap Like
- notification changes into...
- feedback saying that the comment/post was liked
- show error if there was a problem 

I think I like the loading indicator for Reply though. Overall, I'd follow Gmail's lead on the design and UX on these actions. I'd be interested in seeing what Gmail does for replying to an email. Unfortunately, I haven't been able to look at Android 7, and the emulators are taking forever to download :)

Things I noticed in your gif:

- The loading text after you hit reply says "Updating". Can we do something different that is more appropriate?
- Is the dark blue text field standard for this interaction? What do the other Google apps do? Is Gmail a dark red text field?

**Comment by daniloercoli** [comment]
>  I'd be interested in seeing what Gmail does for replying to an email.

No direct reply on Gmail (I tested on my Nexus 5). They open the app and takes you to the standard flow for reply.

**Comment by mzorz** [comment]
just in time @drw158 :D

Ok I can try something for the kind of flow you're describing for "likes" (or "Approves", will be the same)

Regarding these other 2 questions:

> - The loading text after you hit reply says "Updating". Can we do something different that is more appropriate?

Sure, what wording do you think would be better? Maybe "Processing..."? 

> - Is the dark blue text field standard for this interaction? What do the other Google apps do? Is Gmail a dark red text field?

AFAIK this is something I can't change without using a custom view. OTOH, we can use custom views :D, that is, we could even think of a different design altogether. I'm just using stock Android here for this very part.

Let me know what you think, I'd love to make a progressive approach on this (i.e. discussing custom views on a separate PR)

**Comment by daniloercoli** [comment]
Another important thing that I think we should fix asap:
If you have the PIN Lock feature turned on, all of the quick reply actions should be disabled. 
(Or find a way to insert the PIN, but I think this is hard to do in notification center and we also need to modify the passcode library, etc).

@kwonye @jleandroperez  Thoughts?

**Comment by mzorz** [comment]
@daniloercoli re: this

> Another important thing that I think we should fix asap:
> If you have the PIN Lock feature turned on, all of the quick actions should be disabled. 
> 

I don't think we should give this too much thought. Just tried Twitter's FOLLOW quick action on someone how just followed me, and it's enabled and working on Lock screen

File: WordPress/src/main/java/org/wordpress/android/push/GCMMessageService.java [hunk 1]
```diff
@@ -52,15 +53,18 @@
 
 import de.greenrobot.event.EventBus;
 
+import static org.wordpress.android.ui.notifications.NotificationsListFragment.NOTE_INSTANT_REPLY_EXTRA;
+
 public class GCMMessageService extends GcmListenerService {
     private static final ArrayMap<Integer, Bundle> sActiveNotificationsMap = new ArrayMap<>();
+    private static NotificationHelper sNotificationHelpers;
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/push/GCMMessageService.java]: woah, will check this asap

**Comment by mzorz** [comment]
> PIN Lock feature turned on

Sorry, misread your comment and thought of Android lock screen. My bad

File: WordPress/src/main/java/org/wordpress/android/push/GCMMessageService.java (hunk 1 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/push/GCMMessageService.java]: ready here https://github.com/wordpress-mobile/WordPress-Android/pull/4673/commits/6c6eb9ffe5f9b998ae2f71664b8d75b24549b961

**Comment by mzorz** [comment]
This one is ready for merge. Continuing iteration on #4689

**Comment by mzorz** [comment]
Regarding PIN lock treatment, new issue opened here #4690

File: WordPress/src/main/java/org/wordpress/android/push/GCMMessageService.java (hunk 1 above)
Comment (karama78) [inline WordPress/src/main/java/org/wordpress/android/push/GCMMessageService.java]: __Je dois réfléchir j'ais vu le téléphone c'est très compliqué à cause de l'anglais mais je vais réussir

**Comment by daniloercoli** [comment]
:shipit:

**Pull request merged**